### PR TITLE
slice operator supporting arbitrary values of step

### PR DESCRIPTION
**Pull request by reminisce**
## Description ##
- Re-implemented slice operator using Kernel::Launch
- Added support for arbitrary values of `step` along with `begin` and `end`, i.e. `slice(data, begin, end, step)`, where `step` is an optional parameter.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated.
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- The kernel implementation is based upon the parallelization approach of slice operator in mshadow with additional support for arbitrary value of `step`.
- This operator will be used in https://github.com/apache/incubator-mxnet/pull/8246 to support slicing `NDArray` with `step!=1`. Currently, it uses `gather_nd` op to realize that functionality, which has heavy overhead of making index `NDArray` from slices.
- Mini-benchmark `slice_v1` (**mshadow version**) and `slice` (**Kernel::Launch version**).

**Hardware**
p2.xlarge (4 omp threads)

**Commit**
[ba9de66](https://github.com/reminisce/mxnet/commit/ba9de669a752f232dcb21b0c11f9a4f91533af23)

**GPU build**
mx.nd.slice_v1: 10000 repeats costs 0.561281 seconds
mx.nd.slice:      10000 repeats costs 0.562561 seconds

**CPU-only build**
mx.nd.slice_v1: 10000 repeats costs 1.049587 seconds
mx.nd.slice:      10000 repeats costs 1.130866 seconds

**Benchmark script**
```python
 import mxnet as mx
 import numpy as np
 import time
 from mxnet.test_utils import same
 
 #ctx = mx.gpu(0)
 ctx = mx.cpu(0)
 
 repeat = 10000
 shape = (16, 16, 16, 16)
 a = mx.nd.arange(np.prod(shape), ctx=ctx).reshape(shape=shape)
 begin = (None, 1, 2)
 end = (shape[0], shape[1], shape[2])
 
 # warm up
 for i in range(100):
     b = mx.nd.slice_v1(a, begin=begin, end=end)
     c = mx.nd.slice(a, begin=begin, end=end)
 
 mx.nd.waitall()
 start = time.time()
 for i in range(repeat):
     c = mx.nd.slice_v1(a, begin=begin, end=end)
 mx.nd.waitall()
 elapsed = time.time() - start
 print("mx.nd.slice_v1: %d repeats costs %f seconds" % (repeat, elapsed))
 
 start = time.time()
 for i in range(repeat):
     b = mx.nd.slice(a, begin=begin, end=end)
 mx.nd.waitall()
 elapsed = time.time() - start
 print("mx.nd.slice: %d repeats costs %f seconds" % (repeat, elapsed))
 
 assert same(c.asnumpy(), b.asnumpy())
```
@piiswrong @eric-haibin-lin @anirudh2290 @rahul003 @cjolivier01

### CODE REVIEW

File: src/operator/tensor/matrix_op-inl.h [hunk 1]
```diff
@@ -620,6 +626,227 @@ void SliceEx(const nnvm::NodeAttrs& attrs,
...
+      e = param.end[i].value();
+      if (e < 0) {
+        e += len;
+        CHECK_GE(e, 0) << "slicing with end[" << i << "]="
+                       << e - len << " exceeds limit of " << len;
+      }
+    } else if (s < 0) {
+      e = -1;
+    }
+    CHECK_LE(e, len) << "slicing with end[" << i << "]="
```
Comment (cjolivier01) [inline src/operator/tensor/matrix_op-inl.h]: In thsse CHECK_XXX macros, is function call made for every <<, even if the check succeeds?  If so, does this affect performance?

File: src/operator/tensor/matrix_op-inl.h [hunk 2]
```diff
@@ -620,6 +626,227 @@ void SliceEx(const nnvm::NodeAttrs& attrs,
...
+      int stride = 1;
+      int idx = i;
+      #pragma unroll
+      for (int k = ndim - 2; k >= 0; --k) {
+        irow += stride * ((idx % oshape[k]) * step[k] + begin[k]);
+        idx /= oshape[k];
+        stride *= dshape[k];
+      }
+      KERNEL_ASSIGN(igrad[irow*data_last_dim_size+j*step_last_dim+begin_last_dim],
+                    req, ograd[ograd_offset++]);
```
Comment (cjolivier01) [inline src/operator/tensor/matrix_op-inl.h]: nit: spaces around operators

**Comment by cjolivier01** [comment]
For CPU build:
mx.nd.slice_v1: 10000 repeats costs 1.049587 seconds
mx.nd.slice: 10000 repeats costs 1.130866 seconds

Any chance you can break this down into forward and backward times?  Is it known why this is slower than mshadow version?

**Comment by reminisce** [comment]
@cjolivier01 The time is for forward only, no backward (the backward kernel is essentially the same as the forward one). It includes shape/dtype inferences and forward function calls. There might be a few reasons that lead to more runtime in the new implementation:
1. Infer shape function is more complicated than the previous version since we need to support step, while the previous version assumes.
2. Kernel needs to take care of step as well.
3. Kernel needs to copy shapes of data, out, and step in addition to the shapes of begin and end in the mshadow version.
I can investigate to see which part takes this 10% more time.

File: src/operator/tensor/matrix_op.cc [hunk 3]
```diff
@@ -246,11 +246,10 @@ will return a new array with shape ``(2,1,3,4)``.
 .add_argument("data", "NDArray-or-Symbol", "Source input")
 .add_arguments(ExpandDimParam::__FIELDS__());
 
-NNVM_REGISTER_OP(slice)
-.add_alias("_sparse_slice")
-.add_alias("crop")
+NNVM_REGISTER_OP(slice_v1)
```
Comment (piiswrong) [inline src/operator/tensor/matrix_op.cc]: don't need slice_v1

Comment (reminisce) [reply src/operator/tensor/matrix_op.cc]: Removed.

File: src/operator/tensor/matrix_op-inl.h (hunk 1 above)
Comment (reminisce) [reply src/operator/tensor/matrix_op-inl.h]: It's not called if check is successful. See the code:
https://github.com/dmlc/dmlc-core/blob/595d02c0e87be8a0846700462b6f45f1b1031e39/include/dmlc/logging.h#L89
All the << calls happen only when `_check_err` has a non-empty string stored.

**Comment by reminisce** [comment]
@cjolivier01 I tried replacing infer shape and forward kernel with the same ones as in mshadow, but there is still about 10% difference. That means there is some difference between the runs of mshadow and Kernel::Launch. Any insights?

File: src/operator/tensor/matrix_op-inl.h [hunk 4]
```diff
@@ -414,16 +420,6 @@ inline TShape GetSliceShape(const SliceParam& param, const TShape& dshape) {
...
-                       std::vector<TShape> *in_attrs,
-                       std::vector<TShape> *out_attrs) {
-  const TShape& dshape = (*in_attrs)[0];
-  if (dshape.ndim() == 0) return false;
-  const SliceParam& param = nnvm::get<SliceParam>(attrs.parsed);
-  SHAPE_ASSIGN_CHECK(*out_attrs, 0, GetSliceShape(param, dshape));
-  return true;
-}
-
 inline bool SliceForwardInferStorageType(const nnvm::NodeAttrs& attrs,
```
Comment (eric-haibin-lin) [inline src/operator/tensor/matrix_op-inl.h]: Does the slice implementation for CSR support arbitrary step size? Do we want to update the infer storage fucntion to fallback appropriately?

Comment (reminisce) [reply src/operator/tensor/matrix_op-inl.h]: Added fallback check now. Please review again.

**Pull request merged**